Makefile

```diff
@@ -28,10 +28,12 @@ abigen:
 	@echo "Run \"$(BIN)/abigen\" to launch abigen."
 
 test:
-	$(GORUN) build/ci.go test
+	$(GORUN) build/ci.go test -skip ^TestExecutionSpecState$$
```

Comment:
Is this for testing? If so, you can simply run `cd tests; go test -run=TestExecutionSpecState -v`

Reply:
This is for a CI test. Separating tests is needed because `relaxPrecompileRangeForTest` and `isTestExecutionSpecState` affect overall.

Reply:
I think we are not running in parallel, you can run them altogether, of you defer SetRelax...(false)

Reply:
https://github.com/kaiachain/kaia/pull/129#discussion_r1841401749
I couldn't use defer because of this.

Reply:
If we use mutex, can this be reverted?

Reply:
I tested it below and successfully done:

```diff
-  $(GORUN) build/ci.go test -skip ^TestExecutionSpecState$$
- $(GORUN) build/ci.go test -run ^TestExecutionSpecState$$

+ $(GORUN) build/ci.go test    You, 24/11/18 • Uncommitted changes
```

What error leads you to branching with two executions?

Reply:
I don't know why but TestKaiaSpecState failed. Some of the errors is following:

```
--- FAIL: TestKaiaSpecState (2.14s)
    --- FAIL: TestKaiaSpecState/stSolidityTest/TestCryptographicFunctions.json (0.00s)
        --- FAIL: TestKaiaSpecState/stSolidityTest/TestCryptographicFunctions.json/Byzantium/0 (0.00s)
            state_test.go:132: post state root mismatch: got 349489d4e80a0456a178264c7bcf402dba239a4ac6d5aa62b03bb6b645409e6c, want 79f49f0fa14e21191f7513eed2df22be22474565fbe230776b8fc8b46d7e7fd4
            state_test.go:134: gas limit too high for EVM trace
    --- FAIL: TestKaiaSpecState/stReturnDataTest/create_callprecompile_returndatasize.json (0.01s)
        --- FAIL: TestKaiaSpecState/stReturnDataTest/create_callprecompile_returndatasize.json/Byzantium/0 (0.00s)
            state_test.go:132: post state root mismatch: got e988bab3dc9bbbe1eb5d415237f26ed6152a22e83ca77510c77b5b6c5440e34a, want 5546052355a4c86deacb9fe9e0d6ac9f6ff7ae16b3c2900ca87bad7926109d4d
            state_test.go:134: gas limit too high for EVM trace
    --- FAIL: TestKaiaSpecState/stReturnDataTest/call_ecrec_success_empty_then_returndatasize.json (0.00s)
        --- FAIL: TestKaiaSpecState/stReturnDataTest/call_ecrec_success_empty_then_returndatasize.json/Byzantium/0 (0.00s)
            state_test.go:132: post state root mismatch: got 019e9af82544c1354885801c7ad431353abcdd58fdc5aaee141ef6c6d9bccc8a, want 21be96a714bbd3db07c2e3d063b4ed5e3470190b7fa576e45353e83d7f33830e
            state_test.go:134: gas limit too high for EVM trace
```

Reply:
This problem can be solved by removing `t.Parallel()` at L40 in `state_test.go`.

Reply:
@blukat29 `t.Parallel()` seems not much effective and shows no difference in my local. What do you think?

Reply:
Do we need to ensure that other tests are not running in parallel? If so, it's not easy...

Reply:
@hyunsooda 
Can you reproduce it by removing cache?

Reply:
> Do we need to ensure that other tests are not running in parallel? If so, it's not easy...

I belive that `t.Parallel()` is effective in the scope of a specific test. If so, `TestKaiaSpecState` is only the target  of the sequential execution.

> Can you reproduce it by removing cache?

Do you mean `build/cache` directory? If so, I removed it and did `make test` with comment out `t.Parallel()`.

Reply:
> I belive that t.Parallel() is effective in the scope of a specific test. If so, TestKaiaSpecState is only the target of the sequential execution.

When I tried the following code, Test_Func2 could be finished after Test_Func3 started.

```golang
package tests

import (
	"fmt"
	"testing"
)

func trace(name string) func() {
	fmt.Printf("%s entered\n", name)
	return func() {
		fmt.Printf("%s returned\n", name)
	}

}

func Test_Func1(t *testing.T) {
	defer trace("Test_Func1")()

	// ...
}

func Test_Func2(t *testing.T) {
	defer trace("Test_Func2")()
	t.Parallel()

	// ...
}

func Test_Func3(t *testing.T) {
	defer trace("Test_Func3")()

	// ...
}
```

> Do you mean build/cache directory? If so, I removed it and did make test with comment out t.Parallel().

Please use `go clean -testcache`.

Reply:
> Do we need to ensure that other tests are not running in parallel?

I vaguely remember that most tests are not parallel.. We can simply search for `t.Parallel`

Reply:
I'm a little bit confused. If the test is executed in parallel, the function execution order does matter because this PR uses and updates a global variable, but I thought that we were commenting out `t.Parallel()` and expected that `TestExecutionSpecState` and `TestKaiaSpecState` would not corrupt each other. Can you pinpoint the current situation you are encountering so that I can understand more?

Reply:
@hyunsooda 
The problem is running `TestKaiaSpecState` with `relaxPrecompileRangeForTest = true`. This is caused by starting `TestExecutionSpecState` before finishing `TestKaiaSpecState` due to `t.Parallel()`.

I think dropping `t.Parallel()` in all tests and resetting `relaxPrecompileRangeForTest` after finishing `TestExecutionSpecState` are needed so that `TestKaiaSpecState` (`common.RelaxPrecompileRangeForTest(true)`) doesn't affect other tests.

Reply:
I understood the spec of `t.Parallel()`. 

The test pauses when `t.Parallel()` is run and then next test starts. All tests which pause by `t.Parallel()` resume in parallel after all tests are done.

Current problem:
1. run `common.RelaxPrecompileRangeForTest(false)` in `TestKaiaSpecState`
2. run `t.Parallel()` and `TestKaiaSpecState` pauses
3. run `TestExecutionSpecState` and `common.RelaxPrecompileRangeForTest(true)`
4. resume `TestKaiaSpecState` with `relaxPrecompileRangeForTest = true`

Simple solution is running `common.RelaxPrecompileRangeForTest(false)` after `t.Parallel()` in `TestKaiaSpecState`. But the root cause is not running `common.RelaxPrecompileRangeForTest(false)` at the end of `TestExecutionSpecState`. So I'll implement this using Test Suite.

Reply:
Got it, thank you. The other tests seem to have been still omitted in the `Makefile`, don't they?

Reply:
I fixed it at https://github.com/kaiachain/kaia/pull/129/commits/e98fb5f4f5c73976f32baf88f1f1a6e8851eca0b.

Reply:
@blukat29 @ian0371 This thread can be closed as `Makefile` is reverted and remains unchanged.